crates/bevy_pbr/src/meshlet/gpu_scene.rs

```diff
@@ -0,0 +1,871 @@
+            if let Some(material_id) = gpu_scene.material_id_lookup.get(&material_asset_id) {
+                gpu_scene.material_ids_present_in_scene.insert(*material_id);
+                gpu_scene.instance_material_ids.get_mut()[i] = *material_id;
+            }
+        }
+    }
+}
+
+// TODO: Try using Queue::write_buffer_with() in queue_meshlet_mesh_upload() to reduce copies
+fn upload_storage_buffer<T: ShaderSize + bytemuck::Pod>(
+    buffer: &mut StorageBuffer<Vec<T>>,
+    render_device: &RenderDevice,
+    render_queue: &RenderQueue,
+) where
+    Vec<T>: WriteInto,
+{
+    let inner = buffer.buffer();
+    let capacity = inner.map_or(0, |b| b.size());
+    let size = buffer.get().size().get() as BufferAddress;
+
+    if capacity >= size {
+        let inner = inner.unwrap();
+        let bytes = bytemuck::cast_slice(buffer.get().as_slice());
+        render_queue.write_buffer(inner, 0, bytes);
+    } else {
+        buffer.write_buffer(render_device, render_queue);
+    }
+}
+
+pub fn prepare_meshlet_per_frame_resources(
+    mut gpu_scene: ResMut<MeshletGpuScene>,
+    views: Query<(Entity, &ExtractedView, AnyOf<(&Camera3d, &ShadowView)>)>,
+    mut texture_cache: ResMut<TextureCache>,
+    render_queue: Res<RenderQueue>,
+    render_device: Res<RenderDevice>,
+    mut commands: Commands,
+) {
+    gpu_scene
+        .previous_thread_id_starts
+        .retain(|_, (_, active)| *active);
+
+    if gpu_scene.scene_meshlet_count == 0 {
+        return;
+    }
+
+    let gpu_scene = gpu_scene.as_mut();
+
+    gpu_scene
+        .instance_uniforms
+        .write_buffer(&render_device, &render_queue);
+    upload_storage_buffer(
+        &mut gpu_scene.instance_material_ids,
+        &render_device,
+        &render_queue,
+    );
+    upload_storage_buffer(
+        &mut gpu_scene.thread_instance_ids,
+        &render_device,
+        &render_queue,
+    );
+    upload_storage_buffer(
+        &mut gpu_scene.thread_meshlet_ids,
+        &render_device,
+        &render_queue,
+    );
+    upload_storage_buffer(
+        &mut gpu_scene.previous_thread_ids,
+        &render_device,
+        &render_queue,
+    );
+
+    let needed_buffer_size = 4 * gpu_scene.scene_index_count;
+    let visibility_buffer_draw_index_buffer =
+        match &mut gpu_scene.visibility_buffer_draw_index_buffer {
+            Some(buffer) if buffer.size() >= needed_buffer_size => buffer.clone(),
+            slot => {
+                let buffer = render_device.create_buffer(&BufferDescriptor {
+                    label: Some("meshlet_visibility_buffer_draw_index_buffer"),
+                    size: needed_buffer_size,
+                    usage: BufferUsages::STORAGE | BufferUsages::INDEX,
+                    mapped_at_creation: false,
+                });
+                *slot = Some(buffer.clone());
+                buffer
+            }
+        };
+
+    let needed_buffer_size = ((gpu_scene.scene_meshlet_count + 31) / 32) as u64 * 4;
```

Comment:
this could use the newly introduced div_ceil, tested and looks the same
```suggestion
    let needed_buffer_size = gpu_scene.scene_meshlet_count.div_ceil(32) as u64 * 4;
```

Reply:
Good catch, I love that this exists now.